### ISSUE TITLE
Error if downcasting `dyn Any` to another trait object

### ISSUE THREAD

**Issue by alpha-tango-kilo** [rust-lang/rust-clippy#12427]
### What it does

You can't downcast a `dyn Any` to a `dyn OtherTrait`, even if the original concrete type does actually implement `OtherTrait`. This is a footgun that tripped me up recently, as the compiler doesn't say anything, but your downcast always fails (returns `None`). Therefore this should be an error because the code you've written will never do what you want

### Advantage

(no recommended code, but advantages of having the lint)

- Lets users know that this will never work as they want it to
- Saves having to understand how trait objects work to know that this wouldn't work

### Drawbacks

It's causing an error when clippying even though the code would still run fine

### Example

```rust
use core::fmt::Debug;

#[derive(Debug)]
struct Foo;

#[test]
fn it_never_works() {
    use std::any::Any;
    
    let foo: Box<dyn Any> = Box::new(Foo);
    assert!(foo.downcast_ref::<&dyn Debug>().is_none());
    assert!(foo.downcast_ref::<Foo>().is_some());
}
```

[(on playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=77bca7ee4732dea1bb2178a13ef69f54)

Can't be rewritten to work easily

**Comment by alpha-tango-kilo** [comment]
Thanks for the response, that's a fair comment. I guess it'd be difficult to lint for situations where it wouldn't work then, due to the fact you'd need to find the original/underlying type that was cast to `dyn Any`? If so, feel free to close the issue

**Comment by Alexendoo** [comment]
Yeah, it would require some global analysis out of our scope
